packages/core/types/ts3.6/effects.test.ts

```diff
@@ -285,9 +285,6 @@ function* testApply(): SagaIterator {
 function* testCps(): SagaIterator {
   type Cb<R> = (error: any, result: R) => void
 
-  // $ExpectError
```

Comment:
I'm not sure why this is failing in `dtslint`.  I can see the error inside my text editor but even after upgrading dtslint I'm stilling seeing the error.

I felt like the error check itself was relatively harmless so I removed it for now.  I'd be happy to keep digging into this to figure out why it's failing for node v16.

Reply:
Okay so I spent a few hours last night debugging this issue.  `dtslint` downloads `typescript@next` which is targeting [4.7.0-dev.20220301](https://www.npmjs.com/package/typescript/v/4.7.0-dev.20220301).

I'm not exactly sure _why_ but our type tests were relying on an error being expected that is no longer present in 4.x.

This line was not throwing an error in 4.x: https://github.com/redux-saga/redux-saga/blob/master/packages/core/types/effects.test.ts#L289

In 4.x that line was resolving to this overload: https://github.com/redux-saga/redux-saga/blob/master/packages/core/types/effects.d.ts#L516

Prior to 4.x that line would resolve to https://github.com/redux-saga/redux-saga/blob/master/packages/core/types/effects.d.ts#L550 which would throw the error we are expecting.

For whatever reason, in 4.x, they changed the behavior of rest parameters + the usage of `never`.  Instead of `never` I think what we actually wanted was to throw an error.

I saw that you posted in typescript on a proposal to add `throw` in a conditional type and that there are workarounds.  I don't think the code I supplied technically works the way we want it to, but by replace `never` with something that doesn't match the code provided for our test, it fails properly and everything works.

https://github.com/redux-saga/redux-saga/blob/master/packages/core/types/effects.d.ts#L555

Reply:
Thank you for the detailed investigation! I know how time-consuming this stuff can be so I appreciate that you have followed this rabbit hole.

It's somewhat weird that [TS playground on 3.x](https://www.typescriptlang.org/play?ts=3.6.3#code/JYWwDg9gTgLgBAbzgYzAZzgXzgMyhEOAcigFMATAVwA8BaNAQwHMGB6UnHU5GNIgKFCRYiOAGVmDAJIxSUBjGhZc+QiQo16kgf1LVh8HJQB2PYBGMAqOLLQwAwugAUASgBc4yTLkKlCfnBwAJ7ApAA25CjOTgwexpQgAEZyLnAAvAB8iJgu-JhAA) doesn't error for this case. Maybe that's some weirdness of the playground.

To the best of my understanding, [this TS playground](https://www.typescriptlang.org/play?ts=3.6.3#code/JYWwDg9gTgLgBAbzgGQIYGcYBo4CUCmAbvlOvnAL5wBmUEIcA5DAJ5j7oDGUwYMAtDACuYADb5GcAPRS4rdugBccAEwA6dQEYAUNuAA7GCWqpO5AMJh051KNEAjUwGsAPLgB8ibXDgAKEnRQyqj6LDhQHEKiMMq4AJTKhBDAACbecJwhZqIA-L4JcEmp2hS68uQAKqjAoi7IcPgAHkb6KehwISwA2gC6ngC8fr5qI6LKyHFw-Z6dk00tbUMAFsGhOCNqMXAG1CRwFZPTHaGTOftwyvpEJGVs5ACCdgBCQjBomHUNzfit7Z29Azw11I+BcVRqbmBZDq7lht3YcEs6AAYkJ9JwYMAIPoAAqoKCoED4IykFzI-RfBbtYYjfEAcyUx26PUOM1CgPeMBceIJRJJ6DJ+lhlJ+iyRNjsjk4rk67nSZ0eohebwwXJ5hOJJAF5LhPkuwN0KXwnFE+PI1DRGKxFM4VkFIt+fg29MZ-xZUzZLHcvmo+mU5PWtKgDOUSNR6Mx2PVfK1gvcBSKaW0tvQvl8qEuQhA9hIrMQFDiQA) should error (this would replicate the situation when in <4.0 this overload **didn't** match) but for some reason, it doesn't. Perhaps this is dependent on some compilation flags? I've tried turning off some strict settings but it didn't change the outcome.

It's an interesting case that might help me to understand some TS details - so I hope that I will find time this week to test this out locally. 

PS. you can hit `y` on GitHub to create a permalink with the hash in it, so the copied links in comments stay always pointing to the very same code lines (I've already updated links in your comment).

Reply:
I played around with it a little more, I think < 4.0 the rest parameter (the params following (a: number) => {} inside `cps`) was getting evaluated to `[]` which is not assignable to never.  However, in > 4.0 `[]` can be assignable to `never` so it is fitting into that slot.

Regardless, in our types we assumed that for `cps` the `CpsFunctionParameters` will never hit the negative on the conditional type, which is incorrect in > 4.0 which is why it’s being permitted (`[]` can be assigned to `never`).

Reply:
For my own reference, so I don't lose this, here we have a TS playground reproducing the error in 3.6:
[TS playground](https://www.typescriptlang.org/play?ts=3.6.3#code/C4TwDgpgBACgThSA7AJgHgCoFcwBtoQAewEqAzlAIZIgDaAugDRQCCKKpKAfFALxQAKAQH0AXK3admAOlnCAjOOx4IASj49qIdURLlBA2dLFQAlkgBmEOFABKEMllzB1vTTXUB+Ow6fAo4kgQAG7WANwAUBGgkFAAkmQAYuamJJg4+FC6nBRaDMzJSKkQzHGWKSQ8-ADeEVD1WQC2YKDihcWRDVBIAPZIAKLNrQZGJsr4ru7aWcQ5BibmVjbJcGTAMnKKZpbWPmuTVB5Q3gntafZrBRUl8eVFlQHdIeF1DYvX4mUW15EAvrTjAizfQMY5QADkECGIHBj0BMz0KAoAkWu36+EapBcoO86KhWNB2X08O84Pe9wgsPE4N6A2hVKeoTg9CiHAAxrhKAgoDFoMIADKUNbpFQIuZ5JhQAAiEAslD8fEZ1iqUFqXShLRA4hlcr8nQaZHMAHN8EoMkDERRaKibABlHr4PGYpDAehg+2OjFYx5BJn6+qNPymcRCUZmlQHLQ6YFI+biLQbYxbG1QAByukjR28guF6eIPG1svlzn922+FPDmSJsZROxsTqxqhxUAbLp9zzgfwB5rFILdpI1oAZ8OrVq0-YhhqQJspcJ7o8EKdb2Iny8JMYoJIh5OKDPBgecpgZvusLOi4GgOeAIqrG8OdHoKoFQuvgK4kXPsWE9iZZAgN4tcUaHyWAEG+Qhe1jCVFQYFU1QaQctVA2VTEIUtaUGTUQ0MORKzUDR72jS04zLXYVkuKBRmTOs0wzAiozBb8Oz-NA83WZDkHQeAUMIK5VmALgCyVTtXnqHcSHEeCunqfpbFsAB5WxqQAYWoXp-AQX9oGoMtrh5c1wVLLplPkqV+lTFgAFl+mpL5rkBQzRKgX4uxHO9aATSjZAlCdTmuADmBpPpMKHQLxMpITIXpM9eR8LSAMg3JgMfRUmPit8P1ijBKFMXA0H5RL71gxVQ1kXBxH5TNpgXIQAAt4xoRNgHEFMMCqrwoAwdsmSiWKWFwXAACEsGAK98sKiUVR-awWOy3K0Gm1Z-35QSuCicwSDgOU2WgZSwDIVSBoAI0oNkAGsFp4eCBGsOAejgBqQGYBBHGccRbFUcRgh6UwUFeNlqB23BPAET6oG+36Il+XqLygPakiwJA2WAUw+hgLlKExTayDQRIkEKnDpC5I0yEehgqpVMb0bgTGIGx3GkEEwr4cO3ATvOtAtDW+pvH6oaRqpjGsZmhnBNeQIO1ZCAOS5aALER5HUfxtl9oZgmjGJ0mivoCmBFeCwkDaJBGFeDW4BJ8R4cSBWUbRoW6ZFvG1rBiG-oiFWyCEShAiwRojusA5ql+VQgA)

Reply:
I think your analysis is mostly correct but in reality, we hit the negative branch of this conditional in all of those TS versions (take look at the bottom):
[TS playground 3.6](https://www.typescriptlang.org/play?ts=3.6.3#code/C4TwDgpgBACgThSA7AJgHgCoFcwBtoQAewEqAzlAIZIgDaAugDRQCCKKpKAfFALxQAKAQH0AXK3admAOlnCAjOOx4IASj49qIdURLlBA2dLFQAlkgBmEOFABKEMllzB1vTTXUB+Ow6fAo4kgQAG7WANwAUBGgkFAAkmQAYuamJJg4+FC6nBRaDMzJSKkQzHGWKSQ8-ADeEVD1WQC2YKDihcWRDVBIAPZIAKLNrQZGJsr4ru7aWcQ5BibmVjbJcGTAMnKKZpbWPmuTVB5Q3gntafZrBRUl8eVFlQHdIeF1DYvX4mUW15EAvrTjAizfQMY5QADkECGIHBj0BMz0KAoAkWu36+EapBcoO86KhWNB2X08O84Pe9wgsPE4N6A2hVKeoTg9CiHAAxrhKAgoDFoMIADKUNbpFQIuZ5JhQAAiEAslD8fEZ1iqUFqXShLRA4hlcr8nQaZHMAHN8EoMkDERRaKibABlHr4PGYpDAehg+2OjFYx5BJn6+qNPymcRCUZmlQHLQ6YFI+biLQbYxbG1QAByukjR28guF6eIPG1svlzn922+FPDmSJsZROxsTqxqhxUAbLp9zzgfwB5rFILdpI1oAZ8OrVq0-YhhqQJspcJ7o8EKdb2Iny8JMYoJIh5OKDPBgecpgZvusLOi4GgOeAIqrG8OdHoKoFQuvgK4kXPsWE9iZZAgN4tcUaHyWAEG+Qhe1jCVFQYFU1QaQctVA2VTEIUtaUGTUQ0MORKzUDR72jS04zLXYVkuKBRmTOs0wzAiozBb8Oz-NA83WZDkHQeAUMIK5VmALgCyVTtXnqHcSHEeCunqfpbFsAB5WxqQAYWoXp-AQX9oGoMtrh5c1wVLLplPkqV+lTFgAFl+mpL5rkBQzRKgX4uxHO9aATSjZAlCdTmuADmBpPpMKHQLxMpITIXpM9eR8LSAMg3JgMfRUmPit8P1ijBKFMXA0H5RL71gxVQ1kXBxH5TNpgXIQAAt4xoRNgHEFMMCqrwoAwdsmSiWKWFwXAACEsGAK98sKiUVR-awWOy3K0Gm1Z-35QSuCicwSDgOU2WgZSwDIVSBoAI0oNkAGsFp4eCBGsOAejgBqQGYBBHGccRbFUcRgh6UwUFeNlqB23BPAET6oG+36Il+XqLygPakiwJA2WAUw+hgLlKExTayDQRIkEKnDpC5I0yEehgqpVMb0bgTGIGx3GkEEwr4cO3ATvOtAtDW+pvH6oaRqpjGsZmhnBNeQIO1ZCAOS5aALER5HUfxtl9oZgmjGJ0mivoCmBFeCwkDaJBGFeDW4BJ8R4cSBWUbRoW6ZFvG1rBiG-oiNk+jWKADZKyhAiwRojusA5qmh939oEA3VBhr9ny9-hBZp4XVjQXkegsb3GZ4AB6bPugDoO4E-PlhH6O8E5fNBqdp+m04zg2mYXFnKGO06Lq5sF5EeAAGKBc6gbugA)
[TS playground 4.6](https://www.typescriptlang.org/play?ts=4.6.2#code/C4TwDgpgBACgThSA7AJgHgCoFcwBtoQAewEqAzlAIZIgDaAugDRQCCKKpKAfFALxQAKAQH0AXK3admAOlnCAjOOx4IASj49qIdURLlBA2dLFQAlkgBmEOFABKEMllzB1vTTXUB+Ow6fAo4kgQAG7WANwAUBGgkFAAkmQAYuamJJg4+FC6nBRaDMzJSKkQzHGWKSQ8-ADeEVD1WQC2YKDihcWRDVBIAPZIAKLNrQZGJsr4ru7aWcQ5BibmVjbJcGTAMnKKZpbWPmuTVB5Q3gntafZrBRUl8eVFlQHdIeF1DYvX4mUW15EAvrTjAizfQMY5QADkECGIHBj0BMz0KAoAkWu36+EapBcoO86KhWNB2X08O84Pe9wgsPE4N6A2hVKeoTg9CiHAAxrhKAgoDFoMIADKUNbpFQIuZ5JhQAAiEAslD8fEZ1iqUFqXShLRA4hlcr8nQaZHMAHN8EoMkDERRaKibABlHr4PGYpDAehg+2OjFYx5BJn6+qNPymcRCUZmlQHLQ6YFI+biLQbYxbG1QAByukjR28guF6eIPG1svlzn922+FPDmSJsZROxsTqxqhxUAbLp9zzgfwB5rFILdpI1oAZ8OrVq0-YhhqQJspcJ7o8EKdb2Iny8JMYoJIh5OKDPBgecpgZvusLOi4GgOeAIqrG8OdHoKoFQuvgK4kXPsWE9iZZAgN4tcUaHyWAEG+Qhe1jCVFQYFU1QaQctVA2VTEIUtaUGTUQ0MORKzUDR72jS04zLXYVkuKBRmTOs0wzAiozBb8Oz-NA83WZDkHQeAUMIK5VmALgCyVTtXnqHcSHEeCunqfpbFsAB5WxqQAYWoXp-AQX9oGoMtrh5c1wVLLplPkqV+lTFgAFl+mpL5rkBQzRKgX4uxHO9aATSjZAlCdTmuADmBpPpMKHQLxMpITIXpM9eR8LSAMg3JgMfRUmPit8P1ijBKFMXA0H5RL71gxVQ1kXBxH5TNpgXIQAAt4xoRNgHEFMMCqrwoAwdsmSiWKWFwXAACEsGAK98sKiUVR-awWOy3K0Gm1Z-35QSuCicwSDgOU2WgZSwDIVSBoAI0oNkAGsFp4eCBGsOAejgBqQGYBBHGccRbFUcRgh6UwUFeNlqB23BPAET6oG+36Il+XqLygPakiwJA2WAUw+hgLlKExTayDQRIkEKnDpC5I0yEehgqpVMb0bgTGIGx3GkEEwr4cO3ATvOtAtDW+pvH6oaRqpjGsZmhnBNeQIO1ZCAOS5aALER5HUfxtl9oZgmjGJ0mivoCmBFeCwkDaJBGFeDW4BJ8R4cSBWUbRoW6ZFvG1rBiG-oiNk+jWKADZKyhAiwRojusA5qmh939oEA3VBhr9ny9-hBZp4XVjQXkegsb3GZ4AB6bPugDoO4E-PlhH6O8E5fNBqdp+m04zg2mYXFnKGO06Lq5sF5EeAAGKBc6gbugA)

It's also quite understanding that if we adjust the value returned from that "negative" branch then... the error goes away. Even though originally it states this:
```
function cps<(a: number) => void>(fn: (a: number) => void, ...args: never): void
Argument of type '[]' is not assignable to parameter of type 'never'.
```

So this really has to be a change in TS that just starts to treating `never` differently in such situations with spreads.

Technically we could replace that value with `Parameters<Fn>`, as technically the passed in callback doesn't have to actually use the `cpsCallback` that gets injected into it by the saga's runtime. On the other hand... this most likely indicates an error on the user's side, as such effect just wouldn't ever complete, so it's probably better to be pragmatic here than "technically correct". In this very situation this would produce the error on the line that we are discussing but it would allow such a call: 
```
yield cps((a: number) => {}, 42)
```

When I think about it - I don't really see a better way to handle this here than what you've done before: using something that can't be matched. We can't use a good generic constraint as `(...args: any[], last: Foo) => any` is not valid. So we have to "defer" the error to conditional types etc. There is another way than to expect some artificial second argument here though and this could look something like this:
[TS playground](https://www.typescriptlang.org/play?ts=3.6.3#code/C4TwDgpgBACgThSA7AJgHgCoFcwBtoQAewEqAzlAIZIgDaAugDRQCCKKpKAfFALxQAKAQH0AXK3admAOlnCAjOOx4IASj49qIdURLlBA2dLFQAlkgBmEOFABKEMllzB1vTTXUB+Ow6fAo4kgQAG7WANwAUBGgkFAAkmQAYuamJJg4+FC6nBRaDMzJSKkQzHGWKSQ8-ADeEVD1WQC2YKDihcWRDVBIAPZIAKLNrQZGJsr4ru7aWcQ5BibmVjbJcGTAMnKKZpbWPmuTVB5Q3gntafZrBRUl8eVFlQHdIeF1DYvX4mUW15EAvrTjAizfQMY5QADkECGIHBj0BMz0KAoAkWu36+EapBcoO86KhWNB2X08O84Pe9wgsPE4N6A2hVKeoTg9CiHAAxrhKAgoDFoMIADKUNbpFQIuZ5JhQAAiEAslD8fEZ1iqUFqXShLRA4hlcr8nQaZHMAHN8EoMkDERRaKibABlHr4PGYpDAehg+2OjFYx5BJn6+qNPymcRCUZmlQHLQ6YFI+biLQbYxbG1QAByukjR28guF6eIPG1svlzn922+FPDmSJsZROxsTqxqhxUAbLp9zzgfwB5rFILdpI1oAZ8OrVq0-YhhqQJspcJ7o8EKdb2Iny8JMYoJIh5OKDPBgecpgZvusLOi4GgOeAIqrG8OdHoKoFQuvgK4kXPsWE9iZZAgN4tcUaHyWAEG+Qhe1jCVFQYFU1QaQctVA2VTEIUtaUGTUQ0MORKzUDR72jS04zLXYVkuKBRmTOs0wzAiozBb8Oz-NA83WZDkHQeAUMIK5VmALgCyVTtXnqHcSHEeCunqfpbFsAB5WxqQAYWoXp-AQX9oGoMtrh5c1wVLLplPkqV+lTFgAFl+mpL5rkBQzRKgX4uxHO9aATSjZAlCdTmuADmBpPpMKHQLxMpITIXpM9eR8LSAMg3JgMfRUmPit8P1ijBKFMXA0H5RL71gxVQ1kXBxH5TNpgXIQAAt4xoRNgHEFMMCqrwoAwdsmSiWKWFwXAACEsGAK98sKiUVR-awWOy3K0Gm1Z-35QSuCicwSDgOU2WgZSwDIVSBoAI0oNkAGsFp4eCBGsOAejgBqQGYBBHGccRbFUcRgh6UwUFeNlqB23BPAET6oG+36Il+XqLx8ABHLBTAQPaDsoY7TouxIkEKnDpC5I0yEehgqpVMaYC5ShMU2sg0CxwTCpRw7cBO860C0HhvCx7rrFZCAOS5aALCwJA2WAUw+igNl9tp7GaqMfHCaK+gSYEV4LCQd6IARpGIEZtHmYxmWuEYV55bgAnxH6oaRrJimqZmo21rBiG-oiKWyCEShAiwRojusA5ql+VQ3f2z3vd9-2CMD5gABYACZg-dsPuh9v24GYd2QwDoOo6DkOPYEL2U4j9PJf2rPc+z2OE4iIA)

Both solutions are IMHO fine though.

Reply:
Thanks for digging deeper!  How would you like to proceed?  I'm okay with whatever path you think we should take.

Reply:
Let's use my proposed solution - it feels more in the spirit of traditional types~.

Reply:
I've pushed out this change and reverted stuff from [here](https://github.com/redux-saga/redux-saga/pull/2261/files#r825283246) + added a test that I've mentioned there.

We could potentially recheck if other overloads of `cps` don't suffer from some similar issues. Perhaps the `RequireCpsCallback` strategy should be used there too, that doesn't have to be done in this PR but it's also not the worst moment to investigate this.

Reply:
I've also pushed out now a new commit that introduces "cps callback requirement" to all other overloads. I've hacked this around somewhat quickly and I didn't have time to fully validate my changes (although they seem to work as far as I can tell, at least for TS >=3.6). Let's see what CI has to say about this.

Reply:
Awesome, looks great!  How would you like to proceed with this PR?

Reply:
Hm, could u recheck if all CI checks pass now? Im on mobile and its hard to navigate this for me right now. I dont see any checks at all here - perhaps because this is a fork. To fix that we should add `pull_request` as a trigger here:

https://github.com/redux-saga/redux-saga/blob/b854e928514d7d5bdafb9e68fbeb097163ae8ee3/.github/workflows/nodejs.yml#L3

Reply:
Okay I made your suggestion to our github action and pushed it up.  CI is now running for this PR and everything passed.